cdap-master/src/main/java/co/cask/cdap/data/runtime/main/MasterServiceMain.java

```diff
@@ -878,12 +878,21 @@ public void run() {
      * runnable.
      */
     private TwillPreparer prepareExploreContainer(TwillPreparer preparer,
-                                                  Iterable<String> exploreExtraClassPaths) throws IOException {
+                                                  Iterable<String> exploreExtraClassPaths,
+                                                  Iterable<String> yarnAppClassPath) throws IOException {
       // Adding all hive jar files to the container classpath.
       // Ideally we only want it for the explore runnable container, but Twill doesn't support it per runnable.
       // However, setting these extra classpath for non-explore container is fine because those hive will be
       // missing from the container (localization is done by the MasterTwillApplication for the explore runnable only).
-      preparer = preparer.withClassPaths(exploreExtraClassPaths);
+
+      boolean yarnFirst = cConf.getBoolean(Constants.Explore.CONTAINER_YARN_APP_CLASSPATH_FIRST);
+      if (yarnFirst) {
+        // It's ok to have yarn application classpath set here even it can affect non-explore container,
+        // since we anyway have yarn application classpath in the "withApplicationClassPaths".
```

Comment:
I think you should add the JIRA number to the comment, in case someone wants details about why this code is even needed.
Remainder LGTM